lib/jekyll/page.rb

```diff
@@ -183,5 +183,13 @@ def trigger_hooks(hook_name, *args)
     def write?
       true
     end
+
+    private
+    def unescaped_url
+      @url_stash ||= {}
+
+      @unescaped_url = nil unless @url_stash.key?(url)
+      @unescaped_url ||= @url_stash[url] = URL.unescape_path(url)
+    end
```

Comment:
This method is too tricky.

Can you explain to me in English what this method is doing? There must be a better way to go about this.

Reply:
How / why is this method risky?

What this method is doing:
For a straight-forward build session, `url` method returns **a memoized value** computed for the `Page` instance. `unescaped_url` takes that `url` and returns **a memoized unencoded URI** via `Addressable`.
But Parker asked to have the `unescaped_url` *break memoization* when `url` itself breaks memoization via a hook plugin..
So now the `unescaped_url` assigns itself the unencoded uri cached in `url_stash` based on the given `url`..


In English, first check if `url_stash` has a key that equals to the value of `url`. If the key exists, then proceed, otherwise ensure the `unescaped_url` has been reset to `nil`. In the following step, if the value of `unescaped_url` is either `nil` or `false`, then assign it the value from the `url_stash` corresponding to the `url` as key, (the value) being in turn computed by `URL.unescape_path` and the current value of `url`

Reply:
> In English, first check if url_stash has a key that equals to the value of url. If the key exists, then proceed, otherwise ensure the unescaped_url has been reset to nil. In the following step, if the value of unescaped_url is either nil or false, then assign it the value from the url_stash corresponding to the url as key, (the value) being in turn computed by URL.unescape_path and the current value of url

You are speaking Ruby 😏

The method is not _risky_, it is _tricky_; it takes a _lot_ of mental energy to figure out what it is doing.

Is this equivalent?

```ruby
private
def unescaped_url
  return url_stash[url] if url_stash.key?(url)
  url_stash[url] = URL.unescape_path(url)
end

private
def url_stash
  @url_stash ||= {}
end
```

Reply:
yep.. its equivalent..
I had initially implemented something similar, but then Rubocop complained of having too many lines in the class `Jekyll::Document` (305 of 300 IIRC)
Perhaps when we decide to remove the "extra `private` keywords, I can split the method as per your suggestion..

Reply:
@pathawks I've pushed a commit to improve the readability like you suggested.. but I'm using instance variables as I'm unable to even define a private `attr_reader :url_stash` without having Rubocop flag the `Jekyll::Document` class for having too many lines.. `[301/300]`